### PR TITLE
Adds functionality to populate torch generator using torch.thread_safe_generator

### PR DESCRIPTION
**Pull request by divyanshk**
Added thread-safe random number generation to all V2 torchvision random transforms to prevent race conditions when using DataLoader with thread-based workers (worker_method='thread').

This is based on [`torch.thread_safe_generator`](https://github.com/pytorch/pytorch/blob/main/torch/random.py#L237) which returns dataloader thread-worker specific RNG or None otherwise.

### CODE REVIEW

**Review by NicolasHug — COMMENTED** [review]
Thanks for the PR @divyanshk . I think the changes look reasonable.

One thing I'm wondering is how does this affect the multiprocess-based dataloaders? Currently, since TV is using the global torch RNG, that global generator will be [seeded by torch](https://github.com/pytorch/pytorch/blob/6f2866ba70fa2dfd3f7c6c619384005f44cf56ab/torch/utils/data/_utils/worker.py#L276C9-L276C14) using a different seed for each process/worker. This is the correct behavior since we want each worker to have a different RNG.

Is that behavior preserved now that we're using `torch.thread_safe_generator()`?

It'd be good to have tests ensure that's the case (both for multiprocess and multithreaded cases).

**Comment by divyanshk** [comment]
The multiprocessing case remains unchanged because torch.thread_safe_generator will return None for multiprocessing use-case. So for MP, there is no change. Earlier the torch.rand functions received None for generator arg, and now they would get the same. Also added a test case where I confirm the expected behavior for multiprocessing.

File: test/test_transforms_v2.py [hunk 1]
```diff
@@ -8120,3 +8115,64 @@ def test_different_sizes(self, make_input1, make_input2, query):
...
+
+    TRANSFORMS = [
+        transforms.RandomResizedCrop(size=(24, 24)),
+        transforms.RandomRotation(degrees=10),
+        transforms.RandomAffine(degrees=10),
+        transforms.RandomCrop(size=(24, 24), pad_if_needed=True),
+        transforms.RandomPerspective(p=1.0),
+        transforms.RandomErasing(p=1.0),
+        transforms.ScaleJitter(target_size=(24, 24)),
+    ]
```
Comment (NicolasHug) [inline test/test_transforms_v2.py:8049]: We have a few more random transforms in TV that we'll also want to update and test. I think the list you'll find in https://github.com/pytorch/vision/pull/7848/changes should have the proper coverage (but `claude` should be able to find all the relevant ones)

File: test/test_transforms_v2.py [hunk 2]
```diff
@@ -8120,3 +8115,64 @@ def test_different_sizes(self, make_input1, make_input2, query):
...
+        """With multiprocessing DataLoader workers, thread_safe_generator()
+        returns None and transforms use the per-process global RNG.
+        Each worker gets a different seed, so results should differ."""
+        dataset = self.TransformDataset(size=2, transform=transform)
+        dl = DataLoader(dataset, batch_size=1, num_workers=2)
+        batch0, batch1 = list(dl)
+        assert not torch.equal(batch0, batch1)
+
+    @pytest.mark.parametrize("transform", TRANSFORMS, ids=lambda t: type(t).__name__)
+    def test_thread_worker_uses_thread_local_generator(self, transform):
```
Comment (NicolasHug) [inline test/test_transforms_v2.py:8074]: For this multi-threading test, is there a way to test the actual multi-threaded behavior, without the mocking? I.e. ideally I'd like to test the public-facing APIs when a user requests multi-threaded from the DataLoader. I'm not sure what the public entry point is though?

File: test/test_transforms_v2.py (hunk 1 above)
Comment (divyanshk) [reply test/test_transforms_v2.py:8049]: I was missing a lot of random transforms. Updated the PR to cover the transforms in these files in `torchvision/transforms/v2`:
_augment.py
_auto_augment.py
_color.py
_container.py
_geometry.py
_misc.py
_transform.py

This matches the files touched in your PR above. 

RandomHorizontalFlip and RandomVerticalFlip were noisy for the test I want, i.e. two batches are different for different seeded workers. This is because the flipped outputs can be same for different seeds. So I added another reproducibility test which cover the flips and other transforms. This check if two transforms are the same for the same torch.thread_safe_generator value. This should add extra coverage.

File: test/test_transforms_v2.py (hunk 2 above)
Comment (divyanshk) [reply test/test_transforms_v2.py:8074]: I agree. I am using the mocks only because the threading workers isn't landed yet and I want to land this PR first. Once in, I can update these tests. With respect to the transforms we are not simplifying anything though. With the mocking, each transform function is getting a different generator as would happen if done through the dataloading workers.

File: test/test_transforms_v2.py (hunk 1 above)
Comment (NicolasHug) [reply test/test_transforms_v2.py:8049]: Thanks for following up, I think there are 3 remaining transforms that we should try to make thread-safe:

- MixUp and CutMix https://github.com/pytorch/vision/blob/4b142d52241eeb9729f8447a6397e99143f18f3a/torchvision/transforms/v2/_augment.py#L149
- GaussianNoise https://github.com/pytorch/vision/blob/4b142d52241eeb9729f8447a6397e99143f18f3a/torchvision/transforms/v2/functional/_misc.py#L202

I remember from https://github.com/pytorch/vision/pull/7848 that, at the time, there was no way to pass a custom generator to `torch.Distribution`. I'm not sure how to handle that.
